src/microbe_stage/MicrobeAI.cs

```diff
@@ -370,43 +384,86 @@ private void RunAndTumble(Random random)
         microbe.State = Microbe.MicrobeState.Normal;
 
         var usefulCompounds = microbe.TotalAbsorbedCompounds.Where(x => microbe.Compounds.IsUseful(x.Key));
+        var compoundsWeights = ComputeCompoundsSearchWeights(usefulCompounds.ToList());
 
-        // If this microbe lacks glucose, don't bother with ammonia and phosphorous
-        // This algorithm doesn't try to determine if iron and sulfuric acid is useful to this microbe
-        if (microbe.Compounds.GetCompoundAmount(glucose) < 0.5f)
+        float gradientValue = 0.0f;
+        foreach (var compoundWeight in compoundsWeights.ToList())
         {
-            usefulCompounds = usefulCompounds.Where(x => x.Key != ammonia && x.Key != phosphates);
-        }
+            // Note this is about absorbed quantities (which is all microbe has access to), not real ones
+            float compoundDifference = 0.0f;
 
-        float compoundDifference = 0.0f;
-        foreach (var compound in usefulCompounds)
-        {
-            compoundDifference += compound.Value;
+            // No need to add a difference if compound was not absorbed
+            if (microbe.TotalAbsorbedCompounds.ContainsKey(compoundWeight.Key))
+                compoundDifference += microbe.TotalAbsorbedCompounds[compoundWeight.Key];
+
+            // Idem if not absorbed before
+            if (previouslyAbsorbedCompounds.ContainsKey(compoundWeight.Key))
+                compoundDifference -= previouslyAbsorbedCompounds[compoundWeight.Key];
+
+            compoundDifference *= compoundWeight.Value;
+            gradientValue += compoundDifference;
         }
 
+        // Implement a detection threshold to possibly rule out too tiny variations
+        float differenceDetectionThreshold = 0;
```

Comment:
Why is this a variable if this is never changed? Shouldn't this be a constant defined in Constants?

Reply:
At first thought I intended to size it with cell capacity (because 'small' variations depend on your size, after all. But that could indeed be at least related to a constant.